File: V-Sentinel-SDKs/go/vsentinel/detections.go
package vsentinel

import (
	"context"
)

// DetectionsService handles detection-related API calls.
type DetectionsService struct {
	client *Client
}

// DetectionListOptions represents options for listing detections.
type DetectionListOptions struct {
	Query     string
	Severity  DetectionSeverity
	Status    DetectionStatus
	HostID    string
	TimeRange string
	SortBy    string
	SortOrder string
	Limit     int
	Offset    int
}

// toParams converts options to query parameters.
func (o *DetectionListOptions) toParams() map[string]string {
	params := make(map[string]string)

	if o.Query != "" {
		params["query"] = o.Query
	}
	if o.Severity != "" {
		params["severity"] = string(o.Severity)
	}
	if o.Status != "" {
		params["status"] = string(o.Status)
	}
	if o.HostID != "" {
		params["host_id"] = o.HostID
	}
	if o.TimeRange != "" {
		params["time_range"] = o.TimeRange
	}
	if o.SortBy != "" {
		params["sort_by"] = o.SortBy
	}
	if o.SortOrder != "" {
		params["sort_order"] = o.SortOrder
	}
	if o.Limit > 0 {
		params["limit"] = intToString(o.Limit)
	}
	if o.Offset > 0 {
		params["offset"] = intToString(o.Offset)
	}

	return params
}

// List retrieves a list of detections.
func (s *DetectionsService) List(ctx context.Context, opts *DetectionListOptions) (*DetectionList, error) {
	var result DetectionList

	params := make(map[string]string)
	if opts != nil {
		params = opts.toParams()
	}

	if _, ok := params["limit"]; !ok {
		params["limit"] = "50"
	}

	err := s.client.get(ctx, "/detections", params, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Get retrieves a specific detection by ID.
func (s *DetectionsService) Get(ctx context.Context, detectionID string) (*Detection, error) {
	var result Detection

	err := s.client.get(ctx, "/detections/"+detectionID, nil, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Search searches for detections matching a query.
func (s *DetectionsService) Search(ctx context.Context, query string, limit int) (*DetectionList, error) {
	var result DetectionList

	params := map[string]string{
		"query": query,
	}
	if limit > 0 {
		params["limit"] = intToString(limit)
	} else {
		params["limit"] = "50"
	}

	err := s.client.get(ctx, "/detections/search", params, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdateStatusOptions represents options for updating a detection status.
type UpdateStatusOptions struct {
	Status DetectionStatus `json:"status"`
	Notes  string          `json:"notes,omitempty"`
}

// UpdateStatus updates the status of a detection.
func (s *DetectionsService) UpdateStatus(ctx context.Context, detectionID string, opts *UpdateStatusOptions) (*Detection, error) {
	var result Detection

	err := s.client.patch(ctx, "/detections/"+detectionID+"/status", opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Assign assigns a detection to a user.
func (s *DetectionsService) Assign(ctx context.Context, detectionID string, assignee string) (*Detection, error) {
	var result Detection

	body := map[string]string{"assignee": assignee}

	err := s.client.patch(ctx, "/detections/"+detectionID+"/assign", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// AddNote adds a note to a detection.
func (s *DetectionsService) AddNote(ctx context.Context, detectionID string, note string) (*DetectionNote, error) {
	var result DetectionNote

	body := map[string]string{"content": note}

	err := s.client.post(ctx, "/detections/"+detectionID+"/notes", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// ListNotes lists notes for a detection.
func (s *DetectionsService) ListNotes(ctx context.Context, detectionID string, limit int) ([]DetectionNote, error) {
	var result struct {
		Notes []DetectionNote `json:"notes"`
	}

	params := map[string]string{}
	if limit > 0 {
		params["limit"] = intToString(limit)
	}

	err := s.client.get(ctx, "/detections/"+detectionID+"/notes", params, &result)
	if err != nil {
		return nil, err
	}

	return result.Notes, nil
}

File: V-Sentinel-SDKs/typescript/src/services/detections.ts
/**
 * Detections API Service
 *
 * Handles detection-related API operations.
 */

import { AxiosInstance } from 'axios';
import { handleAPIError } from '../client';
import {
  Detection,
  DetectionList,
  DetectionSeverity,
  DetectionStatus,
  DetectionNote,
} from '../models';

/** Options for listing detections */
export interface DetectionListOptions {
  query?: string;
  severity?: DetectionSeverity;
  status?: DetectionStatus;
  hostId?: string;
  timeRange?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

/** Options for updating detection status */
export interface UpdateStatusOptions {
  status: DetectionStatus;
  notes?: string;
}

/** Detections API service */
export class DetectionsService {
  private readonly client: AxiosInstance;
  private readonly basePath = '/detections';

  constructor(client: AxiosInstance) {
    this.client = client;
  }

  /**
   * List detections with optional filtering
   */
  async list(options?: DetectionListOptions): Promise<DetectionList> {
    try {
      const params: Record<string, string | number> = {};

      if (options?.query) params.query = options.query;
      if (options?.severity) params.severity = options.severity;
      if (options?.status) params.status = options.status;
      if (options?.hostId) params.host_id = options.hostId;
      if (options?.timeRange) params.time_range = options.timeRange;
      if (options?.sortBy) params.sort_by = options.sortBy;
      if (options?.sortOrder) params.sort_order = options.sortOrder;
      if (options?.limit) params.limit = options.limit;
      if (options?.offset) params.offset = options.offset;

      if (!params.limit) params.limit = 50;

      const response = await this.client.get<DetectionList>(this.basePath, { params });
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get a specific detection by ID
   */
  async get(detectionId: string): Promise<Detection> {
    try {
      const response = await this.client.get<Detection>(
        `${this.basePath}/${detectionId}`
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Search for detections matching a query
   */
  async search(query: string, limit?: number): Promise<DetectionList> {
    try {
      const params: Record<string, string | number> = { query };
      if (limit) params.limit = limit;
      else params.limit = 50;

      const response = await this.client.get<DetectionList>(
        `${this.basePath}/search`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Update the status of a detection
   */
  async updateStatus(
    detectionId: string,
    options: UpdateStatusOptions
  ): Promise<Detection> {
    try {
      const response = await this.client.patch<Detection>(
        `${this.basePath}/${detectionId}/status`,
        options
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Assign a detection to a user
   */
  async assign(detectionId: string, assignee: string): Promise<Detection> {
    try {
      const response = await this.client.patch<Detection>(
        `${this.basePath}/${detectionId}/assign`,
        { assignee }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Add a note to a detection
   */
  async addNote(detectionId: string, note: string): Promise<DetectionNote> {
    try {
      const response = await this.client.post<DetectionNote>(
        `${this.basePath}/${detectionId}/notes`,
        { content: note }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * List notes for a detection
   */
  async listNotes(detectionId: string, limit?: number): Promise<DetectionNote[]> {
    try {
      const params: Record<string, number> = {};
      if (limit) params.limit = limit;

      const response = await this.client.get<{ notes: DetectionNote[] }>(
        `${this.basePath}/${detectionId}/notes`,
        { params }
      );
      return response.data.notes;
    } catch (error) {
      handleAPIError(error);
    }
  }
}